lightwood/encoders/time_series/rnn.py

```diff
@@ -0,0 +1,178 @@
+
+    def prepare_encoder(self, priming_data, feedback_hoop_function = None):
+        """
+        The usual, run this on the initial training data for the encoder
+        :param priming_data: a list of lists [[time_series], ...]
+        :param feedback_hoop_function: [if you wan to get feedback on the training process]
+        :return:
+        """
+        if self._prepared:
+            raise Exception('You can only call "prepare_encoder" once for a given encoder.')
+
+
+        self._encoder = EncoderRNNNumerical(hidden_size=self._encoded_vector_size)
+        optimizer = optim.Adam(self._encoder.parameters(), lr=self._learning_rate)
+        criterium = nn.MSELoss()
+        self._encoder.train()
+        for i in range(self._train_iters):
+            average_loss = 0
+            total = len(priming_data)
+            for j in range(len(priming_data)):
+                data_tensor = tensor_from_series(priming_data[j])
+                loss = 0
+
+                optimizer.zero_grad()
+                encoder_hidden = self._encoder.initHidden()
+                next_tensor = data_tensor[0]
+                for tensor_i in range(len(priming_data[j])-1):
+                    rand = np.random.randint(2)
+                    # teach from forward as well as from known tensor alteratively
+                    if rand == 1:
+                        next_tensor, encoder_hidden = self._encoder.forward(data_tensor[tensor_i] , encoder_hidden)
+                    else:
+                        next_tensor, encoder_hidden = self._encoder.forward(next_tensor.detach(), encoder_hidden)
+                    loss += criterium(next_tensor, data_tensor[tensor_i+1])
+
+
+                loss = loss
+                average_loss += int(loss)
+                loss.backward()
+                optimizer.step()
+            average_loss = average_loss/total
+            if average_loss < self._stop_on_error:
+                break
+            if feedback_hoop_function is not None:
+                feedback_hoop_function("epoch [{epoch_n}/{total}] average_loss = {average_loss}".format(average_loss=average_loss, epoch_n=i+1, total=self._train_iters))
+            
+        self._prepared = True
+
+    def encode_one(self, data, initial_hidden = None, as_list = False, return_next_value = False):
+        """
+        This method encodes one single row of serial data
+        :param data:  a list of values [1,2,3,4,...]
+        :param initial_hidden: if you want to encode from an initial hidden state other than 0s
+        :param as_list: if you want to return the information as lists
+        :param return_next_value:  if you want to return the next value in the time series too
+
+        :return:  either encoded_value or encoded_value, next_value
+        """
+        if not self._prepared:
+            raise Exception('You need to call "prepare_encoder" before calling "encode" or "decode".')
+
+        self._encoder.eval()
+        with torch.no_grad():
```

Comment:
Ok, so... unrelated to this PR, but:

Initially I thought the `.eval` also stopped auto-grad from working. Reading a bit more on this it seems that the way you are doing it here is "correct" in that eval just puts the model in evaluation mode but doesn't disable autograd.

In this case, should be be using eval+no_grad for every instance when we are just using a model for predicting ? (It's currently not done, unrelated to this PR, but could yield a significant speed gain in many cases I'd wager)

Reply:
sure, do a separate issue, and add that to the other parts needed